test/e2e/network/service.go

```diff
@@ -1852,8 +1853,112 @@ var _ = SIGDescribe("Services", func() {
+	ginkgo.It("should handle load balancer cleanup finalizer for service [Slow] [Feature:ServiceFinalizer]", func() {
+		jig := framework.NewServiceTestJig(cs, "lb-finalizer")
+
+		ginkgo.By("Create load balancer service")
+		svc := jig.CreateTCPServiceOrFail(f.Namespace.Name, func(svc *v1.Service) {
+			svc.Spec.Type = v1.ServiceTypeLoadBalancer
+		})
+
+		defer func() {
+			waitForServiceDeletedWithFinalizer(cs, svc.Namespace, svc.Name)
+		}()
+
+		ginkgo.By("Wait for load balancer to serve traffic")
+		svc = jig.WaitForLoadBalancerOrFail(svc.Namespace, svc.Name, framework.GetServiceLoadBalancerCreationTimeout(cs))
+
+		ginkgo.By("Check if finalizer presents on service with type=LoadBalancer")
+		waitForServiceUpdatedWithFinalizer(cs, svc.Namespace, svc.Name, true)
+
+		ginkgo.By("Check if finalizer is removed on service after changed to type=ClusterIP")
+		jig.ChangeServiceType(svc.Namespace, svc.Name, v1.ServiceTypeClusterIP, framework.GetServiceLoadBalancerCreationTimeout(cs))
+		waitForServiceUpdatedWithFinalizer(cs, svc.Namespace, svc.Name, false)
+
+		ginkgo.By("Check if finalizer is added back to service after changed to type=LoadBalancer")
+		jig.ChangeServiceType(svc.Namespace, svc.Name, v1.ServiceTypeLoadBalancer, framework.GetServiceLoadBalancerCreationTimeout(cs))
+		waitForServiceUpdatedWithFinalizer(cs, svc.Namespace, svc.Name, true)
+	})
 })
 
+func waitForServiceDeletedWithFinalizer(cs clientset.Interface, namespace, name string) {
+	ginkgo.By("Delete service with finalizer")
+	if err := cs.CoreV1().Services(namespace).Delete(name, nil); err != nil {
+		framework.Failf("Failed to delete service %s/%s", namespace, name)
+	}
+
+	ginkgo.By("Wait for service to disappear")
+	if pollErr := wait.PollImmediate(framework.LoadBalancerPollInterval, framework.GetServiceLoadBalancerCreationTimeout(cs), func() (bool, error) {
+		svc, err := cs.CoreV1().Services(namespace).Get(name, metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				e2elog.Logf("Service %s/%s is gone.", namespace, name)
+				return true, nil
+			}
+			return false, err
+		}
+		e2elog.Logf("Service %s/%s still exists with finalizers: %v", namespace, name, svc.Finalizers)
+		return false, nil
+	}); pollErr != nil {
+		framework.Failf("Failed to wait for service to disappear: %v", pollErr)
+	}
+}
+
+func waitForServiceUpdatedWithFinalizer(cs clientset.Interface, namespace, name string, hasFinalizer bool) {
+	ginkgo.By(fmt.Sprintf("Wait for service to hasFinalizer=%t", hasFinalizer))
+	if pollErr := wait.PollImmediate(framework.LoadBalancerPollInterval, framework.GetServiceLoadBalancerCreationTimeout(cs), func() (bool, error) {
+		svc, err := cs.CoreV1().Services(namespace).Get(name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		foundFinalizer := false
+		for _, finalizer := range svc.Finalizers {
+			if finalizer == "service.kubernetes.io/load-balancer-cleanup" {
```

Comment:
use constant from service helper package??

Reply:
oh, I see the constant doesn't exist yet haha sorry

Reply:
Exactly --- I filed https://github.com/kubernetes/kubernetes/issues/78549 to track the followup.